components/api/src/main/java/com/hotels/styx/api/messages/FullHttpRequest.java

```diff
@@ -0,0 +1,694 @@
+     * Builder.
+     *
+     * @param <T> body type
+     */
+    public static final class Builder<T> {
+        private static final InetSocketAddress LOCAL_HOST = createUnresolved("127.0.0.1", 0);
+
+        private Object id;
+        private HttpMethod method = GET;
+        private InetSocketAddress clientAddress = LOCAL_HOST;
+        private boolean validate = true;
+        private Url url;
+        private boolean secure;
+        private final HttpHeaders.Builder headers;
+        private HttpVersion version = HTTP_1_1;
+        private T body;
+        private final List<HttpCookie> cookies;
+
+        public Builder() {
+            this.url = Url.Builder.url("/").build();
+            this.headers = new HttpHeaders.Builder();
+            this.body = null;
+            this.cookies = new ArrayList<>();
+        }
+
+        public Builder(HttpMethod method, String uri) {
+            this();
+            this.method = requireNonNull(method);
+            this.url = Url.Builder.url(uri).build();
+            this.secure = url.isSecure();
+        }
+
+        public Builder(com.hotels.styx.api.HttpRequest request, T body) {
+            this.id = request.id();
+            this.method = request.method();
+            this.clientAddress = request.clientAddress();
+            this.url = request.url();
+            this.secure = request.isSecure();
+            this.version = request.version();
+            this.headers = request.headers().newBuilder();
+            this.body = body;
+            this.cookies = new ArrayList<>(request.cookies());
+        }
+
+        Builder(FullHttpRequest<T> request) {
+            this.id = request.id();
+            this.method = request.method();
+            this.clientAddress = request.clientAddress();
+            this.url = request.url();
+            this.secure = request.isSecure();
+            this.version = request.version();
+            this.headers = request.headers().newBuilder();
+            this.body = request.body();
+            this.cookies = new ArrayList<>(request.cookies());
+        }
+
+        private Builder(FullHttpRequest<?> request, boolean doNotCopyBody) {
+            checkArgument(doNotCopyBody);
+            this.id = request.id();
+            this.method = request.method();
+            this.clientAddress = request.clientAddress();
+            this.url = request.url();
+            this.secure = request.isSecure();
+            this.version = request.version();
+            this.headers = request.headers().newBuilder();
+            this.body = null;
+            this.cookies = new ArrayList<>(request.cookies());
+        }
+
+        /**
+         * Sets the request URI.
+         *
+         * @param uri URI
+         * @return {@code this}
+         */
+        public Builder<T> uri(String uri) {
+            return this.url(Url.Builder.url(uri).build());
+        }
+
+        public Builder<T> body(T content) {
+            setContentLength(content);
```

Comment:
I would remove this for now. For the following reasons:
- The API consumer may want to set `Transfer-Encoding: Chunked` instead.
- For testing purposes, etc we may want to create HTTP messages which *deliberately do not* have content length set. In fact this is a perfectly valid use case for servers returning HttpResponses.
- The content length depends on the kind of encodings applied to the body, and it should be in-line with the other headers applied such as content type charset etc. This method doesn't attempt to set those headers, and also I am not confident the length calculation works in all situations.